### ISSUE TITLE
API documentation

### ISSUE THREAD

**Issue by gitgkk** [zeromq/zeromq.js#127]
I'm new to javascript but have used zeromq before with c and c++. How do I find the names of methods, arguments, options etc here? Where should I look? 

I tried running some examples on windows but it failed to run. Interestingly, the same example runs in Linux. I tried asyncsrv.js from https://github.com/booksbyus/zguide/tree/master/examples/Node.js. 

E:\Workspace\javascript\zmq>node asyncs.js
E:\Workspace\javascript\zmq\node_modules\zeromq\lib\index.js:720
  this._zmq.close();
            ^

TypeError: Socket is closed
    at TypeError (native)
    at Socket.close (E:\gkashyap\Workspace\javascript\zmq\node_modules\zeromq\lib\index.js:720:13)
    at Timeout._onTimeout (E:\gkashyap\Workspace\javascript\zmq\asyncs.js:26:12)
    at ontimeout (timers.js:365:14)
    at tryOnTimeout (timers.js:237:5)
    at Timer.listOnTimeout (timers.js:207:5)
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: write EPIPE
    at exports._errnoException (util.js:1022:11)
    at ChildProcess.target._send (internal/child_process.js:654:20)
    at ChildProcess.target.send (internal/child_process.js:538:19)
    at sendHelper (cluster.js:751:15)
    at send (cluster.js:534:12)
    at exitedAfterDisconnect (cluster.js:467:5)
    at Worker.onmessage (cluster.js:454:7)
    at ChildProcess.<anonymous> (cluster.js:765:8)
    at emitTwo (events.js:111:20)
    at ChildProcess.emit (events.js:191:7)

E:\gkashyap\Workspace\javascript\zmq>node asyncs.js
E:\gkashyap\Workspace\javascript\zmq\node_modules\zeromq\lib\index.js:720
  this._zmq.close();
            ^

TypeError: Socket is closed
    at TypeError (native)
    at Socket.close (E:\gkashyap\Workspace\javascript\zmq\node_modules\zeromq\lib\index.js:720:13)
    at Timeout._onTimeout (E:\gkashyap\Workspace\javascript\zmq\asyncs.js:26:12)
    at ontimeout (timers.js:365:14)
    at tryOnTimeout (timers.js:237:5)
    at Timer.listOnTimeout (timers.js:207:5)
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: write EPIPE
    at exports._errnoException (util.js:1022:11)
    at ChildProcess.target._send (internal/child_process.js:654:20)
    at ChildProcess.target.send (internal/child_process.js:538:19)
    at sendHelper (cluster.js:751:15)
    at send (cluster.js:534:12)
    at exitedAfterDisconnect (cluster.js:467:5)
    at Worker.onmessage (cluster.js:454:7)
    at ChildProcess.<anonymous> (cluster.js:765:8)
    at emitTwo (events.js:111:20)
    at ChildProcess.emit (events.js:191:7)

**Comment by rgbkrk** [comment]
You'll want to use "zeromq" rather than "zmq" (the online guide doesn't match master for zguide yet...).

**Comment by gitgkk** [comment]
Yes, I'm using zeromq. zmq is just the name of my folder. It has nothing to do with the installation.

However, I see this.
E:\gkashyap\Workspace\javascript\zmq>npm -v zeromq
3.10.10

E:\gkashyap\Workspace\javascript\zmq>npm -v zmq
3.10.10

Let us forget for the moment about windows problems. 

I want to know how do I know the API methods for zeromq in javascript library as asked in the original post. Where can I find them? Thanks for your response.

**Comment by rgbkrk** [comment]
We have some autogenerated documentation here but I'll admit that it's fairly lackluster: http://zeromq.github.io/zeromq.js/

Beyond that, there are examples in the README as well as the zguide. If you figure anything out worth documenting, we would _love_ pull requests to add them.

**Comment by gitgkk** [comment]
Well, first of all, as I said, I'm new to javascript and I'm having some trouble understanding its working as I've never programmed with this language. I've seen examples and located the ones which fit my working context and want to use them after some modifications but to do that I need to know what are the methods available to be used and supported by the javascript library. 


Is there a list of methods available?

Could you please help me understand what's happening in these lines of code?

 var args = Array.apply(null, arguments) What does this statement do? Tried to read on the web, I understood it converts the arguments to an array. Is it correct?

sock.send([args[0], '', 'response ' + count++])
I don't understand what is the second argument in this call. Line 58: https://github.com/booksbyus/zguide/blob/master/examples/Node.js/asyncsrv.js

Thanks for your response.

**Comment by gitgkk** [comment]
Nobody to explain the API !

**Comment by jdetle** [comment]
Hi @gitgkk, you are right about `Array.apply`. As for your other questions, all the documented functions are here: https://github.com/zeromq/zeromq.js/blob/master/lib/index.js. If you want to learn more about Javascript, I have enjoyed http://eloquentjavascript.net/. 

As an aside, the people who work on open source projects do this work for free, and their jobs often keep them from having the time to respond to every issue. I understand that it can be frustrating to feel that you don't have adequate resources to use software that you really want to try. However, most developers in OSS are stretched very thin and as much as they would like to help they just don't have the time. (As a new developer myself, this is something I often struggle with.)

Good luck!

**Comment by gitgkk** [comment]
Hi jdetle,
Thanks for the explanation and I did guess so. Just that at the moment I'm not so confident about javascript projects.
I perfectly understand that open source developers are stretched and I've seen so. To join or to start an OSS is a great endeavor and a commitment that must be taken seriously. I've seen some projects where developers left or are looking for somebody else to offload the maintenance of the project as they've other personal or professional commitments. I was searching for a library and I had to remove one project out of my list due to this reason.
Anyway, thanks again for your time and the team for the free lib!
